driver/utils/unicode_converter.h

```diff
@@ -0,0 +1,540 @@
+            const auto * conservative_target_end = (conservative_target_end_candidate < target_end ? conservative_target_end_candidate : target_end);
+            const auto * final_target_end = (trim_pivot_signature && !pivot_signature_trimmed ? conservative_target_end : target_end);
+            UErrorCode error_code = U_ZERO_ERROR;
+
+            ucnv_toUnicode(converter_, &target, final_target_end, &source, source_end, nullptr, true, &error_code);
+
+            target_symbols_written += target - target_prev;
+
+            while (
+                trim_pivot_signature &&
+                !pivot_signature_trimmed &&
+                (target_symbols_written * sizeof(ConverterPivotWideCharType)) >= pivot_signatures_to_trim_max_size_
+            ) {
+                pivot.resize(target_symbols_written); // Shrink to avoid memmove'ing unused bytes.
+                const auto consumed_symbols = consumePivotSignatureInPlace(pivot); // pivot is not shorter than pivot_signatures_to_trim_max_size_, so even 0 is valid.
+                target_symbols_written -= consumed_symbols;
+                target -= consumed_symbols;
+                pivot_signature_trimmed = (consumed_symbols == 0); // This means we may try several times, and stop only when all possible duplicate leading signatures are consumed.
+            }
+
+            if (error_code == U_BUFFER_OVERFLOW_ERROR) {
+                const std::size_t source_pending_bytes = source_end - source;
+                const std::size_t source_pending_symbols = (source_pending_bytes / sizeof(CharType)) + (source_pending_bytes % sizeof(CharType) > 0 ? 1 : 0);
+                const std::size_t target_new_size = target_symbols_written + source_pending_symbols;
+                const std::size_t final_new_size = std::max(target_new_size + extra_size_reserve, pivot.size() + min_size_increment);
+
+                resize_without_initialization(pivot, final_new_size);
+
+                target = const_cast<ConverterPivotWideCharType *>(reinterpret_cast<const ConverterPivotWideCharType *>(pivot.c_str())) + target_symbols_written;
+                target_end = reinterpret_cast<const ConverterPivotWideCharType *>(pivot.c_str() + pivot.size());
+            }
+            else if (U_FAILURE(error_code)) {
+                throw std::runtime_error(u_errorName(error_code));
+            }
+            else {
+                pivot.resize(target_symbols_written);
+                break;
+            }
+        }
+    }
+    catch (...) {
+        ucnv_resetToUnicode(converter_);
+        pivot.clear();
+        throw;
+    }
+}
+
+template <typename CharType, typename PivotCharType>
+inline void UnicodeConverter::convertFromPivot(
+    const std::basic_string_view<PivotCharType> & pivot,
+    std::basic_string<CharType> & encoded,
+    const bool ensure_pivot_signature,
+    const bool trim_encoded_signature
+) {
+    encoded.clear();
```

Comment:
Please add a simplified plan/high-level overview of what is happening here in overly simplified pseudocode or just plain English. Hope it would make the reader's life easier.

Reply:
Done.